### PR TITLE
Add ServerEngine

### PR DESCRIPTION
**Pull request by naritta**
https://github.com/fluent/fluentd/wiki/Fluentd-running-on-ServerEngine

implement
・Logging
・Signal handler
・Daemonize
・Worker processes

### CODE REVIEW

**Comment by repeatedly** [comment]
Could you add remaining list on the description?

**Comment by naritta** [comment]
I got it.

**Comment by naritta** [comment]
remaining is:
・create cache for loading config
・fix rpc_server signal handler
・add tests

**Comment by cosmo0920** [comment]
Here is GitHub. 
We can use task list in descriptions with GFM: https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments

**Comment by naritta** [comment]
Now, the main signal handler in fluentd and serverengine is like below.

||fluentd|serverengine(default)|
|:--:|:--:|:--:|
|INT/TERM|graceful shutdown|graceful shutdown|
|HUP| graceful restart| graceful restart(not when worker_type is "process")|
|USR1|flush buffer|graceful restart|

Then I fixed
・to use INT/TERM and HUP handler on serverengine for graceful shutdown and restart.
・let USR1 handler on fluentd as it is for flush buffer. 

As a result, the signal handler will be same as ordinal fluentd like below.

||fluentd|
|:--:|:--:|
|INT/TERM|graceful shutdown|
|HUP|graceful restart|
|USR1|flush buffer|

**Comment by naritta** [comment]
@repeatedly 
fluentd signal handler(here USR1) is prior to serverengine signal handler, so we need to concern about USR1 restarting function. flush buffer is executed if it got USR1.
But if you concern about it and we should set other signal (like WINCH?) as restarting function on serverengine just in case, please let me know.

**Comment by naritta** [comment]
[MEMO] The remaining problem is
・cause error in binding UDP handle after closing and creating again.
・can't use fiddle in Ruby21-x64 and Ruby22-x64.

File: lib/fluent/command/debug.rb [hunk 1]
```diff
@@ -66,7 +66,11 @@
 
 include Fluent::SystemConfig::Mixin
 
-$log = Fluent::Log.new(STDERR, Fluent::Log::LEVEL_TRACE)
+dl_opts = {}
+dl_opts[:log_level] = Fluent::Log::LEVEL_TRACE
+logdev = Fluent::Test::DummyLogDevice.new
```
Comment (tagomoris) [inline lib/fluent/command/debug.rb]: `Fluent::Test::DummyLogDevice` is just for tests. Use `STDERR` as before.

File: lib/fluent/daemon.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+here = File.dirname(__FILE__)
+$LOAD_PATH << File.expand_path(File.join(here, '..'))
+
+require 'serverengine'
+require 'fluent/supervisor'
+
+server_module = Fluent.const_get(ARGV[0])
+worker_module = Fluent.const_get(ARGV[1])
+# it doesn't call ARGV in block because when reloading config, params will be initialized and then it can't use previous config.
+path = ARGV[2]
```
Comment (tagomoris) [inline lib/fluent/daemon.rb]: `config_path`?

File: lib/fluent/load.rb [hunk 3]
```diff
@@ -17,6 +17,7 @@
 end
 require 'cool.io'
 
+require 'serverengine'
```
Comment (tagomoris) [inline lib/fluent/load.rb]: Please don't add `require` anymore here. Add it on each files really required.

File: lib/fluent/log.rb [hunk 4]
```diff
@@ -53,9 +53,15 @@ def self.str_to_level(log_level_str)
       end
     end
 
-    def initialize(out=STDERR, level=LEVEL_TRACE, opts={})
-      @out = out
-      @level = level
+    def initialize(logger, opts={})
+      # overwrites logger.level= so that config reloading resets level of Fluentd::Log
+      set_logger_orig = logger.class.public_instance_method(:level=).bind(logger)
```
Comment (tagomoris) [inline lib/fluent/log.rb]: Is it not logger, but setter of log level of `logger`, right? If so, use names to mean it.

File: lib/fluent/log.rb [hunk 5]
```diff
@@ -53,9 +53,15 @@ def self.str_to_level(log_level_str)
...
+      me = self
+      logger.define_singleton_method(:level=) {|level| set_logger_orig.call(level); me.level = self.level + 1 }
```
Comment (tagomoris) [inline lib/fluent/log.rb]: It's better to have comments to describe why `me.level = self.level +1`.

File: lib/fluent/plugin/in_forward.rb [hunk 6]
```diff
@@ -57,11 +57,16 @@ def configure(conf)
     def start
       @loop = Coolio::Loop.new
 
-      @lsock = listen
+      socket_manager_path = ENV['SERVERENGINE_SOCKETMANAGER_PATH']
+      if Fluent.windows?
+        socket_manager_path = socket_manager_path.to_i
+      end
+      client = ServerEngine::SocketManager::Client.new(socket_manager_path)
```
Comment (tagomoris) [inline lib/fluent/plugin/in_forward.rb]: It's better to have a utility method for these several lines (it looks to appear very often), but I have a plan of a new plugin helper for tcp/udp/tls sockets. So it's my task following this pull-request.

File: lib/fluent/supervisor.rb [hunk 7]
```diff
@@ -24,8 +24,10 @@
 require 'fluent/plugin'
 require 'fluent/rpc'
 require 'fluent/system_config'
+# require 'serverengine'
+require_relative '../../../serverengine/lib/serverengine'
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: What's this?

File: lib/fluent/supervisor.rb [hunk 8]
```diff
@@ -24,8 +24,10 @@
...
 
-if Fluent.windows?
+if ServerEngine.windows?
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Don't use `ServerEngine.windows?` directly.
If it provides better way to know in windows or not, update `Fluent.windows?` method in fluent/env.rb to return result of `ServerEngine.windows?`.

File: lib/fluent/supervisor.rb [hunk 9]
```diff
@@ -35,23 +37,109 @@
...
-    def self.get_etc_passwd(user)
-      if user.to_i.to_s == user
-        Etc.getpwuid(user.to_i)
-      else
-        Etc.getpwnam(user)
-      end
+  module ServerModule
+    def before_run
+      log_opts = {suppress_repeated_stacktrace: config[:suppress_repeated_stacktrace]}
+      log = Supervisor::LoggerInitializer.new(config[:log_path], config[:log_level], config[:chuser], config[:chgroup], log_opts)
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: It is not `$log`, but `$log` is set to `@logger` at L45.
Is it intentional?

File: lib/fluent/supervisor.rb [hunk 10]
```diff
@@ -35,23 +37,109 @@
...
+
+    def after_start
+      install_supervisor_signal_handlers
+    end
+
+    def install_supervisor_signal_handlers
+      trap :USR1 do
+        $log.debug "fluentd supervisor process get SIGUSR1"
+        supervisor_sigusr1_handler
+      end unless ServerEngine.windows?
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: `Fluent.windows?`

File: lib/fluent/supervisor.rb [hunk 11]
```diff
@@ -35,23 +37,109 @@
...
+    end
+
+    def supervisor_sigusr1_handler
+      # @log.reopen!
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Is it needed?

File: lib/fluent/supervisor.rb [hunk 12]
```diff
@@ -35,23 +37,109 @@
...
+      if pid = @pm.pid
+        Process.kill(:USR1, pid)
+        # don't resuce Erro::ESRSH here (invalid status)
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: "rescue" (it's near from "reuse", so it's better to correct even in easy comment)

File: lib/fluent/supervisor.rb [hunk 13]
```diff
@@ -35,23 +37,109 @@
...
       end
     end
+  end
+
+  def self.load_config(path, params = {})
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Class method for `Fluent` (top-level of Fluentd) namespace is not good practice. That namespace is for methods which is used from anywhere in Fluentd core and plugins.
This `load_config` method looks only for supervisor. So we should use its namespace (e.g., Fluent::Supervisor.load_config).

File: lib/fluent/supervisor.rb [hunk 14]
```diff
@@ -35,23 +37,109 @@
...
+    config_mtime = File.mtime(path)
+    # reuse previous config if last load time is within 5 seconds and mtime of the config file is not changed
+    if Time.now - Time.at(params['pre_loadtime'].to_i) < 5 and config_mtime == params['pre_config_mtime']
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: `if params.has_key?('pre_loadtime') && Time.now - Time.at(...) ...`
I can understand that `Time.at(params['pre_loadtime'].to_i)` means `Time.at(0)` if it's missing, but not so explicit for all. We should write it in more explicit way.

Comment (tagomoris) [inline lib/fluent/supervisor.rb]: And same about `pre_config_mtime`.

File: lib/fluent/supervisor.rb [hunk 15]
```diff
@@ -35,23 +37,109 @@
...
+    inline_config = params['inline_config']
+    if inline_config == '-'
+      config_data << "\n" << STDIN.read
+    elsif inline_config
+      config_data << "\n" << inline_config.gsub("\\n","\n")
+    end
+    fluentd_conf = Fluent::Config.parse(config_data, config_fname, config_basedir, params['use_v1_config'])
+    system_config = SystemConfig.create(fluentd_conf)
+
+    log_level = system_config.log_level || params['log_level']
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Command line specification must be prior than system config specification.

File: lib/fluent/supervisor.rb [hunk 16]
```diff
@@ -35,23 +37,109 @@
...
+    suppress_repeated_stacktrace = system_config.suppress_repeated_stacktrace || params['suppress_repeated_stacktrace']
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: ditto.

File: lib/fluent/supervisor.rb [hunk 17]
```diff
@@ -35,23 +37,109 @@
...
+                                path,
+                                JSON.dump(params)],
+       fluentd: fluentd_conf,
+       main_cmd: main_cmd,
+    }
+    pre_params = params.dup
+    params['pre_loadtime'] = Time.now.to_i
+    params['pre_config_mtime'] = config_mtime
+    params['pre_conf'] = se_config
+    # prevent pre_conf from being too big by reloading many times.
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Very good comment!

File: lib/fluent/supervisor.rb [hunk 18]
```diff
@@ -158,39 +242,11 @@ def start
       set_system_config
 
       dry_run if @dry_run
-      start_daemonize if @daemonize
       setup_rpc_server if @rpc_endpoint
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Following 3 instructions should be in one `if` block.
```
if @rpc_endpoint
  setup_rpc_server
  setup_rpc_get_dump if @enable_get_dump
  run_rpc_server
end
```
Or, these 3 methods might be just one method (if it looks better than 3 methods).

File: lib/fluent/supervisor.rb [hunk 19]
```diff
@@ -203,14 +259,33 @@ def options
       }
     end
 
+    def main_run
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Currently, `start` is for supervised configuration, and `main_run` is for execution without supervisor. But these names don't express it.
Can you consider better names for these methods?

File: lib/fluent/supervisor.rb [hunk 20]
```diff
@@ -243,82 +318,33 @@ def show_plugin_config
...
-        STDIN.reopen("/dev/null")
-        STDOUT.reopen("/dev/null", "w")
-        STDERR.reopen("/dev/null", "w")
-        @wait_daemonize_pipe_w.write @supervisor_pid.to_s
-        @wait_daemonize_pipe_w.close
-        @wait_daemonize_pipe_w = nil
-      end
-    end
-
+    # TODO: move RPC Server in serverengine side
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Is it needed? What for?

File: lib/fluent/supervisor.rb [hunk 21]
```diff
@@ -346,64 +372,73 @@ def stop_rpc_server
...
+        $log.debug "fluentd main process get SIGINT"
+        unless @finished
+          @finished = true
+          $log.debug "getting start to shutdown main process"
+          Fluent::Engine.stop
+        end
       end
 
-      $log.info "process finished", code: ecode
+      trap :USR1 do
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: We want to support force flushing of buffers even in Windows by using RPC server calls, but with this implementation, it can't be done.
This actual procedure should be written in a method, and signal handler (and supervisor-worker RPC handler or so for windows) should call it.

File: lib/fluent/supervisor.rb [hunk 22]
```diff
@@ -243,82 +318,33 @@ def show_plugin_config
...
         $log.debug "fluentd RPC got /api/processes.killWorkers request"
-        supervisor_sigterm_handler
+        Process.kill :TERM, $$
         nil
       }
       @rpc_server.mount_proc('/api/plugins.flushBuffers') { |req, res|
         $log.debug "fluentd RPC got /api/plugins.flushBuffers request"
-        supervisor_sigusr1_handler
+        unless ServerEngine.windows?
+          Process.kill :USR1, $$
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: If ServerEngine provides any supervisor-worker communication path, we should use it instead of signaling SIGUSR1 (because it's not supported by Windows, but we want to provide buffer force flushing even in Windows if possible).

File: lib/fluent/supervisor.rb [hunk 23]
```diff
@@ -243,82 +318,33 @@ def show_plugin_config
...
+        end
         nil
       }
       @rpc_server.mount_proc('/api/config.reload') { |req, res|
         $log.debug "fluentd RPC got /api/config.reload request"
-        $log.info "restarting"
-        supervisor_sighup_handler
+        unless ServerEngine.windows?
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Same with SIGUSR1.

File: lib/fluent/supervisor.rb [hunk 24]
```diff
@@ -346,64 +372,73 @@ def stop_rpc_server
...
-        @wait_daemonize_pipe_w.close
-        @wait_daemonize_pipe_w = nil
+        fluentd_spawn_cmd = $0 + ' '
       end
 
-      Process.waitpid(@main_pid)
-      @main_pid = nil
-      ecode = $?.to_i
+      $fluentdargv.each{|a|
+        fluentd_spawn_cmd << (a + " ")
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Is this cmd string executed vis shell?
If so, these should be escaped with ShellWords. (I don't know well about Windows though.)

File: lib/fluent/supervisor.rb [hunk 25]
```diff
@@ -346,64 +372,73 @@ def stop_rpc_server
...
       end
 
-      $log.info "process finished", code: ecode
+      trap :USR1 do
+        $log.debug "fluentd main process get SIGUSR1"
+        $log.info "force flushing buffered events"
+        @log.reopen!
 
-      if !@finished && Time.now - start_time < 1
-        $log.warn "process died within 1 second. exit."
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Don't these check & warning exist anywhere?

File: lib/fluent/test/base.rb [hunk 26]
```diff
@@ -119,7 +119,11 @@ def close
     class TestLogger < Fluent::PluginLogger
       def initialize
         @logdev = DummyLogDevice.new
-        super(Fluent::Log.new(@logdev))
+        dl_opts = {}
+        dl_opts[:log_level] = Fluent::Log::LEVEL_INFO
+        logger = ServerEngine::DaemonLogger.new(@logdev, dl_opts)
```
Comment (tagomoris) [inline lib/fluent/test/base.rb]: This file looks to need `require 'serverengine'`.

File: test/plugin/test_in_udp.rb [hunk 27]
```diff
@@ -65,37 +77,6 @@ def test_time_format
...
-        tests.each { |test|
-          u.send(test['msg'], 0)
-        }
-        sleep 1
-      end
-
-      compare_test_result(d.emits, tests)
-    end
-  }
-
```
Comment (tagomoris) [inline test/plugin/test_in_udp.rb]: Why are these tests removed?

File: test/test_config.rb [hunk 28]
```diff
@@ -155,16 +155,5 @@ def write_config(path, data)
...
-  def test_inline
-    prepare_config 
-    opts = {
-      config_path: "#{TMP_DIR}/config_test_1.conf",
-      inline_config: "<source>\n  type http\n  port 2222\n </source>"
-    }
-    assert_nothing_raised do
-      Fluent::Supervisor.new(opts)
-    end  
-  end
```
Comment (tagomoris) [inline test/test_config.rb]: ditto.

**Comment by tagomoris** [comment]
@naritta I added comments for code. Please check it out.

**Comment by naritta** [comment]
Thank you very much.
I will check it.

File: lib/fluent/supervisor.rb (hunk 9 above)
Comment (naritta) [inline lib/fluent/supervisor.rb]: ```$log``` is initialized in ```log.init``` and pass this ```$log``` to serverengine.
If it cause confusion, I'll change the name.

File: lib/fluent/supervisor.rb (hunk 15 above)
Comment (naritta) [inline lib/fluent/supervisor.rb]: I think we can reflect new setting by reloading config file when dynamic reloading.
If it's wrong, please tell me.

File: lib/fluent/supervisor.rb (hunk 22 above)
Comment (naritta) [inline lib/fluent/supervisor.rb]: About reloading config by rpc, I fixed to kill worker and restart by auto restarting function on windows, as temporary plan.

About flushing buffer by rpc, I think we have to implement supervisor-worker communication path is serverengine.
If you have any better idea, please tell me.

**Comment by naritta** [comment]
@tagomoris 
I fixed and added some replying comments.
Could you review when you have time?

File: lib/fluent/log.rb [hunk 29]
```diff
@@ -53,9 +53,17 @@ def self.str_to_level(log_level_str)
...
 
-    def initialize(out=STDERR, level=LEVEL_TRACE, opts={})
-      @out = out
-      @level = level
+    def initialize(logger, opts={})
+      # overwrites logger.level= so that config reloading resets level of Fluentd::Log
+      orig_logger_level_setter = logger.class.public_instance_method(:level=).bind(logger)
+      me = self
+      # DaemonLogger's log level setting has difference of one level from that of this class.
+      # Then level in this class will be self.level + 1.
```
Comment (tagomoris) [inline lib/fluent/log.rb]: Does DaemonLogger have same # of levels with Fluentd's? (It's just difference about origin 0 or 1?)
If so, good to note it.

File: lib/fluent/supervisor.rb (hunk 9 above)
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Did you remove these lines?

Comment (naritta) [inline lib/fluent/supervisor.rb]: I moved these lines into load_config for when reloading. Sorry for not telling.

Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Ah, I found it. Thanks.

File: lib/fluent/supervisor.rb [hunk 30]
```diff
@@ -35,21 +37,187 @@
...
+      log_level = system_config.log_level || params['log_level']
+      suppress_repeated_stacktrace = system_config.suppress_repeated_stacktrace || params['suppress_repeated_stacktrace']
+      log_path = params['log_path']
+      chuser = params['chuser']
+      chgroup = params['chgroup']
+      rpc_endpoint = system_config.rpc_endpoint
+      enable_get_dump = system_config.enable_get_dump
+
+      log_opts = {suppress_repeated_stacktrace: suppress_repeated_stacktrace}
+      log = Supervisor::LoggerInitializer.new(log_path, log_level, chuser, chgroup, log_opts)
```
Comment (tagomoris) [inline lib/fluent/supervisor.rb]: Finally I thought that variable name `log` is bad name. There're many `log` used as logger (and almost same about `$log`), but this `log` is not logger, but initializer.
So this name should be another one to show "this is initializer, not logger itself", and it's better to have comment to show that `.init` sets initialized logger to `$log`.

File: test/plugin/test_in_udp.rb [hunk 31]
```diff
@@ -63,23 +80,33 @@ def test_time_format
...
+          {'msg' => {'k' => 123, 'message' => 'tcptest1'}.to_json + "\n", 'expected' => 'tcptest1'},
+          {'msg' => {'k' => 'tcptest2', 'message' => 456}.to_json + "\n", 'expected' => 456},
+      ],
+      '/^\\[(?<time>[^\\]]*)\\] (?<message>.*)/' => [
+          {'msg' => '[Sep 10 00:00:00] localhost: ' + 'x' * 100 + "\n", 'expected' => 'localhost: ' + 'x' * 100},
+          {'msg' => '[Sep 10 00:00:00] localhost: ' + 'x' * 1024 + "\n", 'expected' => 'localhost: ' + 'x' * 1024},
+      ]
   }.each { |format, test_cases|
     define_method("test_msg_size_#{format[0] == '/' ? 'regexp' : format}") do
+      if Fluent.windows?
```
Comment (tagomoris) [inline test/plugin/test_in_udp.rb]: Doesn't `ServerEngine::SocketManager::Server` support Windows environment?

**Comment by tagomoris** [comment]
I added some more review comments right now.
@naritta Could you comment how you will fix (or why you won't fix these) per review comments? It's a bit hard to follow how they are fixed from entire diff, and becomes much harder if code are moved to another place.

File: lib/fluent/supervisor.rb (hunk 30 above)
Comment (naritta) [inline lib/fluent/supervisor.rb]: I fixed here.
https://github.com/fluent/fluentd/pull/880/commits/6ee3397f90066ea332b63cb38dd05ca8dcf22034

File: test/plugin/test_in_udp.rb (hunk 31 above)
Comment (naritta) [inline test/plugin/test_in_udp.rb]: I thought it's safer that we use original unused_port method in fluentd.
I fixed to use serverengine's api here. https://github.com/fluent/fluentd/pull/880/commits/75a07d24881a8087d6d0ea070b774869f562baf4

File: lib/fluent/log.rb (hunk 29 above)
Comment (naritta) [inline lib/fluent/log.rb]: The comment is not written in serverengine side, because serverengine is used for excepting fluentd also.
If this answer is not your hoping, please tell me.

**Comment by naritta** [comment]
@tagomoris 
Thank you for reviewing and sorry for inconvenience.
I fixed and added comments.
And I also added test for supervisor.

File: lib/fluent/log.rb (hunk 29 above)
Comment (tagomoris) [inline lib/fluent/log.rb]: I had a doubt that:
* `level` of Fluentd logger is a number, between 0-5
* the `+1` means ServerEngine's logger level is also a number, but between 1-6.
* is it "really" collect?

Fluentd engineer might be unfamiliar to ServerEngine (because these are different product :)), so we should add comments about ServerEngine's log levels especially for the case like doing `+1` on Fluentd's log level.

File: test/plugin/test_in_forward.rb [hunk 32]
```diff
@@ -7,6 +7,22 @@
 require 'fluent/plugin/in_forward'
 
 class ForwardInputTest < Test::Unit::TestCase
+  class << self
+    def startup
+      if Fluent.windows?
+        socket_manager_path = ServerEngine::SocketManager::Server.generate_path
+      else
+        socket_manager_path = ServerEngine::SocketManager::Server.generate_path + __method__.to_s
+      end
```
Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: Could you reply here about what the difference between windows and else?
In other words, I want to know the difference between `Server.generate_path` and `Server.generate_path + __method__.to_s` and why (to write plugin helper for these sockets).

Comment (naritta) [inline test/plugin/test_in_forward.rb]: serverengine's API generates
・in unix
generates Unix domain socket and path is time + pid
・in windows
generates TCPServer and path is unused port.

Then, in unix path will be same when the api is called at same time and in same process.
(In windows, it will be not same bacause it will find in each time when it's called)
So I added ``` __method__.to_s``` not to be same in unix.

But It seems to be confusing and I realized we need to create only one socket manager server before tests begin. (we don't have to create for each one test)
Then I fixed it now. Sorry for taking your time.

File: lib/fluent/log.rb (hunk 29 above)
Comment (naritta) [inline lib/fluent/log.rb]: I added the comment.
https://github.com/fluent/fluentd/pull/880/commits/3cb29cb9912052e715c0972701923906a4ee973a

This is for that Serverengine logger ```level=``` overwrites fluentd's logger level. (Not fluentd logger ```level=``` overwrites serverengine)

**Comment by tagomoris** [comment]
This looks ready to merge.

**Comment by tagomoris** [comment]
Done!

**Comment by naritta** [comment]
Thank you!
